File: 08-stackQueues/important/zig/LruCache.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
var gpa = std.heap.GeneralPurposeAllocator(.{}).init;

fn Node(comptime T: type) type {
    return struct {
        const Self = @This();
        key: usize,
        value: *T,
        next: ?*Self,
        prev: ?*Self,

        pub fn init(allocator: Allocator, key: usize, value: ?T) !*Self {
            var node = try allocator.create(Self);
            const new_value = try allocator.create(T);
            if (value != null) {
                new_value.* = value.?;
            }

            node.key = key;
            node.value = new_value;
            node.next = null;
            node.prev = null;
            return node;
        }

        pub fn deinit(node: *Self, allocator: Allocator) void {
            allocator.destroy(node.value);
            allocator.destroy(node);
        }
    };
}

fn LruCache(comptime T: type) type {
    const TNode = Node(T);
    const NodeMap = std.AutoHashMap(usize, *TNode);
    return struct {
        const Self = @This();
        capacity: usize,
        mpp: NodeMap,
        head: *TNode,
        tail: *TNode,
        allocator: Allocator,

        pub fn init(allocator: Allocator, _cap: usize) !*Self {
            const lrucache = try allocator.create(Self);
            lrucache.allocator = allocator;
            lrucache.capacity = _cap;
            lrucache.head = try TNode.init(allocator, 0, null);
            lrucache.tail = try TNode.init(allocator, 0, null);
            lrucache.mpp = NodeMap.init(allocator);
            lrucache.head.next = lrucache.tail;
            lrucache.tail.prev = lrucache.head;
            return lrucache;
        }

        pub fn deinit(self: *Self) void {
            var currentNode: *TNode = self.head;
            var nextNode: *TNode = undefined;
            while (currentNode.next != null) {
                nextNode = currentNode.next.?;
                currentNode.deinit(currentNode, self.allocator);
                currentNode = nextNode;
            }
            self.mpp.deinit();
            self.allocator.destroy(self);
        }

        pub fn addNode(self: *Self, node: *TNode) void {
            var temp: *TNode = self.head.next.?;
            node.next = temp;
            node.prev = self.head;
            self.head.next = node;
            temp.prev = node;
        }

        pub fn deleteNode(self: *Self, node: *TNode) void {
            var delprev = node.prev;
            var delnext = node.next;
            delprev.?.next = delnext;
            delnext.?.prev = delprev;
            node.deinit(self.allocator);
        }

        pub fn get(self: *Self, _key: usize) !?T {
            if (self.mpp.contains(_key)) {
                const temp = self.mpp.get(_key).?;
                const res: T = temp.value.*;
                _ = self.mpp.remove(_key);
                self.deleteNode(temp);
                self.addNode(temp);
                _ = try self.mpp.put(_key, self.head.next.?);
                return res;
            }
            return null;
        }

        pub fn put(self: *Self, _key: usize, value: T) !void {
            if (self.mpp.contains(_key)) {
                const existingNode = self.mpp.get(_key).?;
                _ = self.mpp.remove(_key);
                self.deleteNode(existingNode);
            }
            if (self.mpp.count() == self.capacity) {
                _ = self.mpp.remove(self.tail.prev.?.key);
                self.deleteNode(self.tail.prev.?);
            }
            self.addNode(try TNode.init(self.allocator, _key, value));
            try self.mpp.put(_key, self.head.next.?);
        }

        pub fn print(self: *Self) void {
            var temp: ?*TNode = self.head.next.?;
            while (temp.?.next != null) {
                std.debug.print("{d} : {s}\n", .{ temp.?.key, temp.?.value.* });
                temp = temp.?.next;
            }
        }
    };
}

pub fn main() !void {
    const lruCacheString = LruCache([]const u8);
    const allocator: Allocator = gpa.allocator();
    var lrucache = try lruCacheString.init(allocator, 2);
    try lrucache.put(1, "one");
    try lrucache.put(2, "two");
    try lrucache.put(3, "three");
    try lrucache.put(4, "four");
    try lrucache.put(5, "five");
    const temp = try lrucache.get(3);
    if (temp != null) {
        std.debug.print("{s}\n", .{temp.?});
    }
    lrucache.print();
}

File: 08-stackQueues/important/LruCache.cpp
#include "LruCache.h"

template <typename T> LruCache<T>::LruCache(int cap) {
    capacity = cap;
    head->next = tail;
    tail->prev = head;
}

template <typename T> void LruCache<T>::add(Node<T> *n) {
    Node<T> *temp = this->head->next;
    n->next = temp;
    n->prev = head;
    this->head->next = n;
    temp->prev = n;
}

template <typename T> void LruCache<T>::deleteNode(Node<T> *n) {
    Node<T> *delprev = n->prev;
    Node<T> *delNext = n->next;
    delprev->next = delNext;
    delNext->prev = delprev;
}

template <typename T> T LruCache<T>::get(int _key) {
    if (this->mpp.find(_key) != this->mpp.end()) {
        Node<T> *temp = this->mpp[_key];
        T res = temp->value;
        this->mpp.erase(_key);
        this->deleteNode(temp);
        this->add(temp);
        this->mpp[_key] = this->head->next;
        return res;
    }
    return T();
}

template <typename T> LruCache<T>::~LruCache() {
    Node<T> *current = head;
    Node<T> *nextNode;
    while (current != nullptr) {
        nextNode = current->next;
        delete current;
        current = nextNode;
    }
}

template <typename T> void LruCache<T>::put(int _key, T value) {
    if (mpp.find(_key) != mpp.end()) {
        Node<T> *existingNode = mpp[_key];
        mpp.erase(_key);
        deleteNode(existingNode);
    }
    if (mpp.size() == capacity) {
        mpp.erase(tail->prev->key);
        deleteNode(tail->prev);
    }
    add(new Node<T>(_key, value));
    mpp[_key] = head->next;
}

int main(int argc, char const *argv[]) {
    LruCache<string> *pp = new LruCache<string>(5);
    pp->put(1, "one");
    pp->put(2, "two");
    pp->put(3, "three");
    pp->put(4, "four");
    pp->put(5, "five");
    cout << pp->get(1) << endl;
    pp->put(6, "six");
    cout << pp->get(2) << endl;
    cout << pp->get(3) << endl;
    cout << pp->get(4) << endl;
    cout << pp->get(5) << endl;
    cout << pp->get(6) << endl;
    delete pp;
    return 0;
}
